runtime/src/iree/base/queue.c

```diff
@@ -0,0 +1,73 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/base/queue.h"
+
+#include "iree/base/assert.h"
+
+void iree_queue_initialize(iree_queue_t* queue,
+                                     iree_allocator_t allocator,
+                                     iree_host_size_t element_size,
+                                     iree_host_size_t inline_count) {
+  queue->allocator = allocator;
+  queue->elements = &queue->initial_allocation[0];
+  queue->element_size = element_size;
+  queue->element_count = 0;
+  queue->capacity = inline_count;
+  queue->head = 0;
+}
+
+void iree_queue_deinitialize(iree_queue_t* queue) {
+  IREE_ASSERT_ARGUMENT(queue);
+  if (queue->elements != &queue->initial_allocation[0]) {
+    iree_allocator_free(queue->allocator, queue->elements);
+  }
+}
+
+iree_status_t iree_queue_push_back(iree_queue_t* queue,
+                                             void* element) {
+  // Expand
+  if (queue->capacity == queue->element_count) {
+    uint8_t* new_mem;
+    queue->capacity = iree_max(16, queue->capacity << 1);
+    IREE_RETURN_IF_ERROR(iree_allocator_malloc(
```

Comment:
you can realloc if this is not the inline allocation to potentially give the allocator a better chance of doing an accounting trick - may nto matter here for its usage but is good habit (won't help a realloc of 8 to 18888, but will likely help 32 to 64 or 1024 to 1025, etc). reallocs also let you avoid the copy or only do partial memsets instead of touching everything.

Reply:
I had originally chosen intentionally to avoid the realloc because of the additional complexity involved, but have changed now.